### PR TITLE
Implement builtin assert! macro

### PR DESCRIPTION
**Pull request by edwin0cheng**
This PR add a dummy implementation for `assert!` macro, which mainly make `hover` and `goto-def` works on arguments inside it.

### CODE REVIEW

**Comment by edwin0cheng** [comment]
OT: Do anyone know why `assert` is a builtin macro, but `assert_eq` doesn't ?

File: crates/ra_hir_expand/src/builtin_macro.rs [hunk 1]
```diff
@@ -493,6 +533,22 @@ mod tests {
         assert_eq!(expanded, "\"\"");
     }
 
+    #[test]
+    fn test_assert_expand() {
+        let expanded = expand_builtin_macro(
+            r#"
+            #[rustc_builtin_macro]
+            macro_rules! assert {
+                ($fmt:expr) => ({ /* compiler built-in */ });
```
Comment (bjorn3) [inline crates/ra_hir_expand/src/builtin_macro.rs]:
```suggestion
                ($cond:expr) => ({ /* compiler built-in */ });
```

File: crates/ra_hir_expand/src/builtin_macro.rs [hunk 2]
```diff
@@ -493,6 +533,22 @@ mod tests {
...
+                ($fmt:expr, $($args:tt)*) => ({ /* compiler built-in */ })
```
Comment (bjorn3) [inline crates/ra_hir_expand/src/builtin_macro.rs]:
```suggestion
                ($cond:expr, $($args:tt)*) => ({ /* compiler built-in */ })
```

File: crates/ra_hir_expand/src/builtin_macro.rs [hunk 3]
```diff
@@ -151,6 +152,45 @@ fn column_expand(
     Ok(expanded)
 }
 
+fn assert_expand(
+    _db: &dyn AstDatabase,
+    _id: LazyMacroId,
+    tt: &tt::Subtree,
+) -> Result<tt::Subtree, mbe::ExpandError> {
+    // A hacky implementation for goto def and hover
+    // We expand `assert!("", arg1, arg2)` to
```
Comment (bjorn3) [inline crates/ra_hir_expand/src/builtin_macro.rs]:
```suggestion
    // We expand `assert!(cond, arg1, arg2)` to
```

File: crates/ra_hir_expand/src/builtin_macro.rs [hunk 4]
```diff
@@ -151,6 +152,45 @@ fn column_expand(
...
+    // ```
+    // {(&(arg1), &(arg2));}
```
Comment (bjorn3) [inline crates/ra_hir_expand/src/builtin_macro.rs]: Please expand to:

```suggestion
    // {(cond, &(arg1), &(arg2));}
```

**Comment by bjorn3** [comment]
> OT: Do anyone know why assert is a builtin macro, but assert_eq doesn't ?

I believe the plan was to show the evaluation of the individual sub-expressions of the condition, but this never got implemented. The RFC did get accepted I think though.

**Pull request merged**